File: perl/Problem10.pl
#!/usr/bin/env perl

package Problem10;

use strict;
use warnings;
use constant {
    UNKNOWN => 0,
    NONPRIME => 1
};

sub sieve {
    my $limit = shift;
    if ($limit < 2) {
        return ();
    }
    my @search = (UNKNOWN) x $limit;
    $search[0] = NONPRIME;
    $search[1] = NONPRIME;

    my $sqrt_limit = sqrt($limit);
    for (my $current = 2; $current <= $sqrt_limit; $current++) {
        for (my $i = $current * 2; $i < $limit; $i += $current) {
            $search[$i] = NONPRIME;
        }
    }
    my @res;
    for (my $i = 0; $i < $limit; $i++) {
        if ($search[$i] == UNKNOWN) {
            push(@res, $i);
        }
    }
    return @res;
}

package main;

use strict;
use warnings;
use List::Util qw/sum/;

print(sum(Problem10::sieve(2e6)));

File: php/Problem10.php
<?php
/*
 * The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.
 * Find the sum of all the primes below two million.
 */

function sieve($limit) {
    if ($limit < 2) {
        return [];
    }
    $search = array_fill(2, $limit - 2, true);

    $sqrt_limit = sqrt($limit);
    for ($current = 2; $current <= $sqrt_limit; $current++) {
        for ($i = $current * 2; $i < $limit; $i += $current) {
            unset($search[$i]);
        }
    }
    return array_keys($search);
}

echo array_sum(sieve(2000000));
